### ISSUE TITLE
Inconsistent behavior of batch normalization layer

### ISSUE THREAD

**Issue by kofd** [fchollet/keras#270]
As you can see by the code below, the behavior of batch normalization seems to be dependent on the batch of samples. Any given sample should always be classified exactly the same (in mode 0 at least). Note that for the single item batch, the output of model.predict(arr1[:1]) is identical to the output of model.predict(arr1[1:]).

```
>>> import numpy
>>> 
>>> from keras.models import Sequential
>>> from keras.layers.core import Dense, Activation
>>> from keras.layers.normalization import BatchNormalization
>>> from keras.optimizers import SGD
>>> 
>>> model = Sequential()
>>> model.add(Dense(200,5))
>>> model.add(BatchNormalization(5))
>>> model.add(Activation('softmax'))
>>> sgd = SGD(lr=0.01, decay=3e-6, momentum=0.9, nesterov=False)
>>> model.compile(loss='categorical_crossentropy', optimizer=sgd)
>>> arr1 = numpy.random.rand(2,200)
>>> y1 = numpy.random.rand(2,5)
>>> model.train(arr1,y1)
array(3.693775110217432)
>>> model.predict(arr1)
2/2 [==============================] - 0s
array([[ 0.21000414,  0.20690492,  0.19180691,  0.19093657,  0.20034746],
       [ 0.1894262 ,  0.19338509,  0.20825628,  0.20882832,  0.2001041 ]])
>>> model.predict(arr1[:1])
1/1 [==============================] - 0s
array([[ 0.19959944,  0.20018073,  0.20001223,  0.1998318 ,  0.2003758 ]])
>>> model.predict(arr1[1:])
1/1 [==============================] - 0s
array([[ 0.19959944,  0.20018073,  0.20001223,  0.1998318 ,  0.2003758 ]])
```

I've read the source code a thousand times and cannot find a reason for this to happen. I've got a working theory that there is an issue with running mean and running std being mutable within get_output(). Those values are also not inspectable as they are not stored as a tensor value:
```
>>> import theano
>>> theano.printing.debugprint(model.layers[1].running_mean)
Elemwise{true_div,no_inplace} [@A] ''   
 |Sum{axis=[0], acc_dtype=float64} [@B] ''   
 | |Elemwise{add,no_inplace} [@C] ''   
 |   |dot [@D] ''   
 |   | |<TensorType(float64, matrix)> [@E]
 |   | |<TensorType(float64, matrix)> [@F]
 |   |DimShuffle{x,0} [@G] ''   
 |     |<TensorType(float64, vector)> [@H]
 |DimShuffle{x} [@I] ''   
   |Subtensor{int64} [@J] ''   
     |Elemwise{Cast{float64}} [@K] ''   
     | |Shape [@L] ''   
     |   |Elemwise{add,no_inplace} [@C] ''   
     |Constant{0} [@M]
```

This issue makes batch normalization fairly useless to me as I cannot guarantee that a sample will always generate the same prediction.

**Comment by fchollet** [comment]
> I've got a working theory that there is an issue with running mean and running std being mutable within get_output(). Those values are also not inspectable as they are not stored as a tensor value

So what would be your suggestion for a fix? Put these quantities inside Theano tensors?

**Comment by alexatbanjo** [comment]
It appears that trying to mutate the running mean and std during training time is considered destructive, and according to theano's documentation should be handled by optimizers. I'm trying to figure out exactly what this means. I'm trying to figure out what this means for how the operation needs to be implemented, but I think this might only solve the issue I'm having with inspection. I keep looking over the math for computing the batch normalization result and the logic really does seem sound. The only thing I would suggest is to use T.std(X, axis=0) instead of T.mean((X-m)**2 + self.epsilon, axis=0) ** 0.5. The resulting mean and std should definitely have the correct dimensions, which means that by all logic this issue shouldn't be happening at all even with the mutation occurring within the function.
